### PR TITLE
[DPC-4420] add pagination to GET /v1/Patient

### PR DESCRIPTION
**Pull request by lukey-luke**
## 🎫 Ticket

[DPC-4420](https://jira.cms.gov/browse/DPC-4420)

## 🛠 Changes

 - added optional query parameters: `_count` and `_page`
 - these are used to paginate results of `patientSearch()`

## ℹ️ Context

 - this is a continuation from PoC that was conducted in [DPC-4175](https://jira.cms.gov/browse/DPC-4175)
 - This aims to address an issue where customers with too many patients to return will have this request fail and cause a degradation of service.
   - Legacy behavior is still enabled, meaning the full roster will be output when `_page` is not provided

## 🧪 Validation

 - Unit tests in `PatientResourceUnitTest.java`

### CODE REVIEW

File: dpc-api/src/test/java/gov/cms/dpc/api/resources/v1/PatientResourceUnitTest.java [hunk 1]
```diff
@@ -120,6 +122,190 @@ public void testPatientSearchNoIdentifier() {
...
+        assertNull(emptyBundle.getLink("next"));
+        assertEquals(emptyBundle.getLink("self").getUrl(), requestPath + hugePageNumber);
+        assertNotNull(actualResponse);
+        assertTrue(actualResponse.getEntry().isEmpty(), "Expected no entries for an out-of-bounds page");
+
+        assertEquals(1, actualResponse.getTotal());
+    }
+
+    @Test
+    public void testPageNumEqualsZero() {
```
[the remark this answers is not in the document]
Comment (lukey-luke) [reply dpc-api/src/test/java/gov/cms/dpc/api/resources/v1/PatientResourceUnitTest.java:172]: apologies, this is incorrect. Count=0 should be the same as summary account. Link to FHIR docs [HERE](https://build.fhir.org/search.html#_count)

> If _count has the value 0 (zero), this SHALL be treated the same as _summary=count

File: dpc-api/src/main/java/gov/cms/dpc/api/resources/v1/PatientResource.java [hunk 2]
```diff
@@ -104,12 +109,24 @@ public Bundle patientSearch(@ApiParam(hidden = true)
...
-        return request.execute();
+        if (page >= 0) {
+            return PagingUtils.handlePaging(request, count, page, "/v1/Patient");
+        }
+        else {
+            return request.execute(); // deprecated - legacy behavior for clients relying on full roster
+        }
+    }
+
+    public Bundle patientSearch(OrganizationPrincipal organization, String patientMBI, int page) {
```
[the remark this answers is not in the document]
Comment (lukey-luke) [reply dpc-api/src/main/java/gov/cms/dpc/api/resources/v1/PatientResource.java]: These overloaded versions were helpers for unit testing. There's no reason for them to live here. Moved to `PatientResourceUnitTest.java`

File: dpc-api/src/main/java/gov/cms/dpc/api/resources/v1/PatientResource.java [hunk 3]
```diff
@@ -110,12 +102,42 @@ public Bundle patientSearch(@ApiParam(hidden = true)
...
+    @Override
+    public Bundle patientSearch(@ApiParam(hidden = true)
+                                @Auth OrganizationPrincipal organization,
+                                @ApiParam(value = "Patient MBI")
+                                @QueryParam(value = Patient.SP_IDENTIFIER) @NoHtml String patientMBI,
+                                @ApiParam(value = "Patients per page")
+                                @QueryParam(value = "_count") Integer count,
+                                @ApiParam(value = "Page number") // null means "do not paginate" for compatibility reasons
+                                @QueryParam(value = "_page") Integer page) {
+        var request = this.buildPatientSearchQuery(organization.getOrganization().getId(), patientMBI);
```
[the remark this answers is not in the document]
Comment (jdettmannnava) [reply dpc-api/src/main/java/gov/cms/dpc/api/resources/v1/PatientResource.java:127]: According to the spec, we can use anything we want to make forward and back work. I think using _offset would work best, as it is what attribution expects.

[the remark this answers is not in the document]
Comment (jdettmannnava) [reply dpc-api/src/main/java/gov/cms/dpc/api/resources/v1/PatientResource.java:127]: You are absolutely correct that the _page logic would have met the spec of the ticket, so good on that! But it kind of fights with _count, which I think is better. @MEspositoE14s , do you have thoughts?

File: dpc-api/src/main/java/gov/cms/dpc/api/resources/v1/PatientResource.java [hunk 4]
```diff
@@ -110,12 +102,42 @@ public Bundle patientSearch(@ApiParam(hidden = true)
...
+        if (count == null && page != null) {
+            count = this.defaultPageSize;
+        }
+
+        if (count != null && count == 0) {
+            request.count(0).summaryMode(SummaryEnum.COUNT); // count gets omitted in request
+            return request.execute();
+        }
+        else if (count !=null && page != null && page >= 1) {
```
Comment (MEspositoE14s) [inline dpc-api/src/main/java/gov/cms/dpc/api/resources/v1/PatientResource.java:136]: If they pass us a negative `_count` or `_page` can/should we throw an error and return a 4xx?

**Review by MEspositoE14s — COMMENTED** [review]
Can you add integration tests at the API level that prove everything works together the way we're expecting? 

1. One happy path with paging.
2. One where they ask for a summary.
3. One where they send an invalid count/page.
4. One where patients are added/deleted in between pages being downloaded.
5. Maybe one where there's exactly one page of patients to be sure it handles that edge case.

**Comment by lukey-luke** [comment]
Note: after further inspection, it looks like an absolute URL path is required in order to retain compatibility with `.next()` calls in PatientResource test e.g. `Bundle secondPage = client.loadPage().next(firstPage).execute();` [HERE](https://github.com/CMSgov/dpc-app/blob/55fcc7129bf8ba2c31de62d11e97a026ffb45a9a/dpc-api/src/test/java/gov/cms/dpc/api/resources/v1/PatientResourceTest.java#L696).

I added a new provider to attribution in order to get the dpc-api url from attribution where links are generated. `String proveBaseURL()` [HERE](https://github.com/CMSgov/dpc-app/blob/bee4c1b7bad966706b8236472f2708570addbfc0/dpc-attribution/src/main/java/gov/cms/dpc/attribution/AttributionAppModule.java#L78C1-L79C1)


I resolved all the threads related to page count as we're now using `_offset` as discussed above.

**Comment by lukey-luke** [comment]
I've updated existing tests to reflect use of `_offset` instead of a page number. I've also added additional tests for `PatientDAO` and integration tests in `dpc-api/src/test/java/gov/cms/dpc/api/resources/v1/PatientResourceTest.java`.

When you have time, please have another look and let me know what you think! @jdettmannnava @MEspositoE14s !!

File: dpc-api/src/main/java/gov/cms/dpc/api/resources/v1/PatientResource.java [hunk 5]
```diff
@@ -336,4 +339,26 @@ private static Optional<WebApplicationException> validateAndAddOrg(Patient patie
...
+
+    private IQuery<Bundle> buildPatientSearchQuery(String orgId, @Nullable String patientMBI) {
+        IQuery<Bundle> query = this.client
+                .search()
+                .forResource(Patient.class)
+                .encodedJson()
+                .where(Patient.ORGANIZATION.hasId(orgId))
+                .returnBundle(Bundle.class);
+        if (patientMBI != null && !patientMBI.isEmpty()) {
+            // Handle MBI parsing
```
Comment (jdettmannnava) [inline dpc-api/src/main/java/gov/cms/dpc/api/resources/v1/PatientResource.java:351]: I think these comments can disappear. DPC-432 no longer exists.

File: dpc-attribution/src/main/java/gov/cms/dpc/attribution/resources/v1/PatientResource.java [hunk 6]
```diff
@@ -33,44 +37,71 @@ public class PatientResource extends AbstractPatientResource {
...
+            daoSearchQuery.setPatientMBI(patientIdentifier.getValue());
+        }
+
+        if (summary != null && summary.equals("count")) {
+            daoSearchQuery.setCount(0);
+            int totalPatients = this.dao.countMatchingPatients(daoSearchQuery);
+            return this.pagingService.convertToSummaryBundle(totalPatients);
+        }
+        if (count == null) {
+            // Legacy behavior - before _page parameter was introduced
```
Comment (jdettmannnava) [inline dpc-attribution/src/main/java/gov/cms/dpc/attribution/resources/v1/PatientResource.java:85]: _count

File: dpc-attribution/src/main/resources/application.yml [hunk 7]
```diff
@@ -38,6 +38,9 @@ server:
             version: ${APP_VERSION:-"unknown_version"}
             application: ${APPLICATION:-"dpc-attribution"}
 
+# The root URL at which the application is accessible, if necessary, include the port, do not include the application version
+publicURL: ${PUBLIC_URL:-"http://localhost:3002"}
```
Comment (jdettmannnava) [inline dpc-attribution/src/main/resources/application.yml:42]: Will need dpc-ops update

**Review by jdettmannnava — COMMENTED** [review]
Should we be including total in the paged response?

**Comment by lukey-luke** [comment]
> Should we be including total in the paged response?

@jdettmannnava I specifically avoided included it in the response, because AFAIK this would require an additional database query. My thinking was that given the nature of the issue, focus on performance. If there's a client demand and we can make the concession, definitely possible

File: dpc-api/src/main/java/gov/cms/dpc/api/resources/v1/PatientResource.java [hunk 8]
```diff
@@ -336,4 +344,46 @@ private static Optional<WebApplicationException> validateAndAddOrg(Patient patie
...
+                expandedMBI = String.format("%s|%s", DPCIdentifierSystem.MBI.getSystem(), patientMBI);
+            }
+            query = query.where(Patient.IDENTIFIER.exactly().identifier(expandedMBI));
+        }
+
+        return query;
+    }
+
+    private void handlePagingLinkPathing(Bundle bundle) {
+        if (bundle == null || bundle.getLink() == null) {
```
Comment (MEspositoE14s) [inline dpc-api/src/main/java/gov/cms/dpc/api/resources/v1/PatientResource.java:369]:
```suggestion
        if (bundle == null || bundle.hasLink() {
```

HAPI has a bunch of methods like this you can use to simplify null and empty checks.

File: dpc-api/src/main/java/gov/cms/dpc/api/resources/v1/PatientResource.java [hunk 9]
```diff
@@ -336,4 +344,46 @@ private static Optional<WebApplicationException> validateAndAddOrg(Patient patie
...
+            return;
+        }
+
+        for (Bundle.BundleLinkComponent link : bundle.getLink()) {
+            String relation = link.getRelation();
+            if (relation == null) {
```
Comment (MEspositoE14s) [inline dpc-api/src/main/java/gov/cms/dpc/api/resources/v1/PatientResource.java:375]: Same as above:

```suggestion
            if (link.hasRelation()) {
```

File: dpc-attribution/src/main/resources/application.yml (hunk 7 above)
Comment (lukey-luke) [reply dpc-attribution/src/main/resources/application.yml]: The domain part of the URL is now handled in [handlePagingLinkPathing() in dpc-api](https://github.com/CMSgov/dpc-app/blob/a91e0bcd60eace4dfb67c123fc6e9e3cbd3db78f/dpc-api/src/main/java/gov/cms/dpc/api/resources/v1/PatientResource.java#L368), removed!

File: dpc-api/src/main/java/gov/cms/dpc/api/resources/v1/PatientResource.java (hunk 8 above)
Comment (lukey-luke) [reply dpc-api/src/main/java/gov/cms/dpc/api/resources/v1/PatientResource.java]: updated to use `!bundle.hasLink()`

**Review by MEspositoE14s — APPROVED** [review]
In hindsight, we probably should've broken this up into multiple tickets, but good job getting through it all!

**Review by jdettmannnava — APPROVED** [review]
LGTM
